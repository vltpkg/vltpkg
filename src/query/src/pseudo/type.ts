import { splitDepID } from '@vltpkg/dep-id/browser'
import type { ParserState } from '../types.ts'
import { asPostcssNodeWithChildren, asTagNode } from '../types.ts'
import { removeDanglingEdges, removeNode } from './helpers.ts'

/**
 * :type(str) Pseudo-Element will match only nodes that are of
 * the same type as the value used. The type is determined by the
 * first part of the dependency ID.
 */
export const type = async (state: ParserState) => {
  const type = asPostcssNodeWithChildren(state.current)
  const selector = asPostcssNodeWithChildren(type.nodes[0])
  const name = asTagNode(selector.nodes[0]).value
  for (const node of state.partial.nodes) {
    const nodeType = splitDepID(node.id)[0]
    if (nodeType !== name) {
      removeNode(state, node)
    }
  }
  removeDanglingEdges(state)
  return state
}
