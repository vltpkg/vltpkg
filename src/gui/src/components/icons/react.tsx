export const React = (props: React.SVGProps<SVGSVGElement>) => {
  return (
    <svg
      width="500"
      height="500"
      viewBox="0 0 500 500"
      fill="none"
      xmlns="http://www.w3.org/2000/svg"
      {...props}>
      <g clip-path="url(#clip0_2967_809)">
        <path
          d="m250.504 284.862c270.046 284.862 285.889 269.029 285.889 249.499c285.889 229.968 270.046 214.136 250.504 214.136c230.961 214.136 215.119 229.968 215.119 249.499c215.119 269.029 230.961 284.862 250.504 284.862z"
          fill="currentColor"
        />
        <path
          d="m250.501 321.96c355.363 321.96 440.371 289.522 440.371 249.509c440.371 209.495 355.363 177.058 250.501 177.058c145.639 177.058 60.6318 209.495 60.6318 249.509c60.6318 289.522 145.639 321.96 250.501 321.96z"
          stroke="currentColor"
          stroke-width="18.913"
        />
        <path
          d="m187.724 285.725c240.155 376.482 310.768 433.837 345.442 413.83c380.116 393.823 365.721 304.031 313.29 213.274c260.859 122.517 190.247 65.1621 155.572 85.1689c120.898 105.176 135.293 194.968 187.724 285.725z"
          stroke="currentColor"
          stroke-width="18.913"
        />
        <path
          d="m187.726 213.277c135.295 304.034 120.9 393.826 155.574 413.833c190.248 433.84 260.861 376.485 313.292 285.728c365.723 194.971 380.118 105.179 345.443 85.172c310.769 65.1652 240.157 122.52 187.726 213.277z"
          stroke="currentColor"
          stroke-width="18.913"
        />
      </g>
      <defs>
        <clipPath id="clip0_2967_809">
          <rect
            width="397"
            height="353"
            fill="currentColor"
            transform="translate(52 73)"
          />
        </clipPath>
      </defs>
    </svg>
  )
}
