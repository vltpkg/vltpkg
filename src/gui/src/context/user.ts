import { createContext } from 'react-router'

import type { User } from '@/state/auth.ts'

/**
 * Router context for authenticated user
 * Used by auth middleware to pass user data to loaders/components
 */
export const userContext = createContext<User | null>(null)
