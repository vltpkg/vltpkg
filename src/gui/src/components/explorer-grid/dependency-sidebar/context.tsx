import { createContext, useEffect, useContext, useRef } from 'react'
import { createStore, useStore } from 'zustand'
import { useToast } from '@/components/hooks/use-toast.ts'
import { useNavigate } from 'react-router'
import { useGraphStore } from '@/state/index.ts'
import { addRemoveDependency } from '@/lib/add-remove-dependency.ts'
import { deepEqual } from '@/utils/deep-equal.ts'

import type { StoreApi } from 'zustand'
import type { PropsWithChildren } from 'react'
import type { GridItemData } from '@/components/explorer-grid/types.ts'
import type { DependencySideBarProps } from '@/components/explorer-grid/dependency-sidebar/index.tsx'
import type { Operation } from '@/lib/add-remove-dependency.ts'
import type { Filter } from './filter-config.ts'

type DependencySidebarState = {
  importerId?: DependencySideBarProps['importerId']
  dependencies: DependencySideBarProps['dependencies']
  filteredDependencies: DependencySideBarProps['dependencies']
  filters: Filter[]
  searchTerm: string
  addedDependencies: string[]
  uninstalledDependencies: DependencySideBarProps['uninstalledDependencies']
  inProgress: boolean
  error?: string
  dependencyPopoverOpen?: boolean
}
type DependencySidebarAction = {
  setAddedDependencies: (
    deps: DependencySidebarState['addedDependencies'],
  ) => void
  setFilteredDependencies: (
    deps: DependencySidebarState['filteredDependencies'],
  ) => void
  setFilters: (filters: DependencySidebarState['filters']) => void
  setSearchTerm: (
    search: DependencySidebarState['searchTerm'],
  ) => void
  setInProgress: (bool: DependencySidebarState['inProgress']) => void
  onDependencyClick: (item: GridItemData) => () => undefined
  setError: (str: DependencySidebarState['error']) => void
  setDependencyPopoverOpen: (
    o: DependencySidebarState['dependencyPopoverOpen'],
  ) => void
}

export type DependencySidebarStore = DependencySidebarState &
  DependencySidebarAction

const DependencySidebarContext = createContext<
  StoreApi<DependencySidebarStore> | undefined
>(undefined)

type DependencySidebarProviderProps = PropsWithChildren & {
  dependencies: DependencySidebarState['dependencies']
  importerId?: DependencySidebarState['importerId']
  onDependencyClick: DependencySidebarAction['onDependencyClick']
  uninstalledDependencies: DependencySidebarState['uninstalledDependencies']
}

/**
 * Deeply compares incoming props and updates the store only if values have changed.
 * Prevents unnecessary store updates and downstream rerenders.
 */
const useSyncStoreProps = <T,>(
  store: StoreApi<T>,
  props: Partial<T>,
): void => {
  const prevPropsRef = useRef<Partial<T>>({})

  useEffect(() => {
    const prev = prevPropsRef.current
    const changedKeys = Object.keys(props).filter(key => {
      const typedKey = key as keyof T
      return !deepEqual(props[typedKey], prev[typedKey])
    })

    if (changedKeys.length > 0) {
      const updated: Partial<T> = {}
      changedKeys.forEach(key => {
        const typedKey = key as keyof T
        updated[typedKey] = props[typedKey]
      })
      store.setState(updated)
      prevPropsRef.current = props
    }
  }, [props, store])
}

export const DependencySidebarProvider = ({
  dependencies,
  uninstalledDependencies,
  importerId,
  onDependencyClick,
  children,
}: DependencySidebarProviderProps) => {
  const query = useGraphStore(state => state.query)
  const dependencySidebarStore = useRef(
    createStore<DependencySidebarStore>(set => ({
      dependencies,
      filteredDependencies: dependencies,
      uninstalledDependencies,
      importerId,
      filters: [],
      searchTerm: '',
      onDependencyClick,
      addedDependencies: [],
      inProgress: false,
      error: undefined,
      dependencyPopoverOpen: false,
      setAddedDependencies: (
        deps: DependencySidebarState['addedDependencies'],
      ) => set(() => ({ addedDependencies: deps })),
      setFilters: (filters: DependencySidebarState['filters']) =>
        set(() => ({ filters })),
      setSearchTerm: (search: string) =>
        set(() => ({ searchTerm: search })),
      setFilteredDependencies: (
        filteredDependencies: DependencySidebarState['filteredDependencies'],
      ) => set(() => ({ filteredDependencies })),
      setInProgress: (bool: DependencySidebarState['inProgress']) =>
        set(() => ({ inProgress: bool })),
      setError: (str: DependencySidebarState['error']) =>
        set(() => ({ error: str })),
      setDependencyPopoverOpen: (
        o: DependencySidebarState['dependencyPopoverOpen'],
      ) => set(() => ({ dependencyPopoverOpen: o })),
    })),
  ).current

  /** The `filteredDependencies` by default will the dependencies themselves */
  useEffect(() => {
    dependencySidebarStore.setState(() => ({
      filteredDependencies: dependencies,
    }))
  }, [dependencies, dependencySidebarStore])

  /** Reset the filters when navigating */
  useEffect(() => {
    dependencySidebarStore.setState(() => ({
      filters: [],
    }))
  }, [query, dependencySidebarStore])

  useSyncStoreProps(dependencySidebarStore, {
    dependencies,
    uninstalledDependencies,
    importerId,
    onDependencyClick,
  })

  return (
    <DependencySidebarContext.Provider value={dependencySidebarStore}>
      {children}
    </DependencySidebarContext.Provider>
  )
}

export const useDependencySidebarStore = <T,>(
  selector: (state: DependencySidebarStore) => T,
) => {
  const store = useContext(DependencySidebarContext)
  if (!store) {
    throw new Error(
      'useDependencySidebarStore must be used within a DependencySidebarProvider',
    )
  }
  return useStore(store, selector)
}

/**
 * usePopover is scoped to the DependencySidebarProvider, and is intended
 * for controlling the state of the `addedDependenciesPopover` component.
 *
 * It will throw an error if used outside of the DependencySidebarProvider.
 */
export const usePopover = () => {
  const setInProgress = useDependencySidebarStore(
    state => state.setInProgress,
  )
  const error = useDependencySidebarStore(state => state.error)
  const setError = useDependencySidebarStore(state => state.setError)

  const dependencyPopoverOpen = useDependencySidebarStore(
    state => state.dependencyPopoverOpen,
  )
  const setDependencyPopoverOpen = useDependencySidebarStore(
    state => state.setDependencyPopoverOpen,
  )

  const toggleAddDepPopover = () => {
    if (dependencyPopoverOpen && error) {
      setError('')
      setInProgress(false)
    }
    setDependencyPopoverOpen(!dependencyPopoverOpen)
  }

  return {
    toggleAddDepPopover,
    dependencyPopoverOpen,
    setDependencyPopoverOpen,
  }
}

export const useOperation = () => {
  const navigate = useNavigate()
  const { toast } = useToast()
  const { toggleAddDepPopover } = usePopover()

  const updateStamp = useGraphStore(state => state.updateStamp)
  const updateErrorCause = useGraphStore(
    state => state.updateErrorCause,
  )

  const importerId = useDependencySidebarStore(
    state => state.importerId,
  )
  const setError = useDependencySidebarStore(state => state.setError)
  const setInProgress = useDependencySidebarStore(
    state => state.setInProgress,
  )
  const addedDependencies = useDependencySidebarStore(
    state => state.addedDependencies,
  )
  const setAddedDependencies = useDependencySidebarStore(
    state => state.setAddedDependencies,
  )

  const operation = async ({
    item,
    operationType,
  }: {
    item: {
      name: GridItemData['name']
      version: GridItemData['version']
      type?: string
    }
    operationType: Operation
  }) => {
    if (!importerId) return

    const onSuccessful =
      operationType === 'install' ?
        () => {
          toggleAddDepPopover()
          setAddedDependencies([item.name, ...addedDependencies])
        }
      : () => {
          const packageName = item.name
          setAddedDependencies(
            addedDependencies.filter(item => item !== packageName),
          )
        }

    try {
      await addRemoveDependency({
        operation: operationType,
        setError,
        setInProgress,
        updateStamp,
        toast,
        importerId,
        name: item.name,
        version: item.version,
        type: item.type,
        onSuccessful,
      })
    } catch (err) {
      console.error(err)
      void navigate('/error')
      updateErrorCause(
        'Unexpected error trying to install dependency.',
      )
    }
  }

  return { operation }
}
