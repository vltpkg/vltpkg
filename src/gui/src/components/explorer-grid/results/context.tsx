import { useEffect, useRef, useContext, createContext } from 'react'
import { useStore, createStore } from 'zustand'
import {
  useNavigate,
  useSearchParams,
  useLocation,
} from 'react-router'
import { PAGE_SIZE_OPTIONS } from '@/components/explorer-grid/results/page-options.tsx'

import type { PropsWithChildren } from 'react'
import type { StoreApi } from 'zustand'
import type { PageSizeOption } from '@/components/explorer-grid/results/page-options.tsx'
import type { GridItemData } from '@/components/explorer-grid/types.tsx'

export type ResultsSortBy =
  | 'alphabetical'
  | 'version'
  | 'dependencyType'
  | 'dependents'
  | 'moduleType'
  | 'overallScore'

type ResultsStoreState = {
  /**
   * The current page number
   */
  page: number
  /**
   * The total number of pages
   */
  totalPages: number
  /**
   * The count of items on each page
   */
  pageSize: PageSizeOption
  /**
   * The items to display on the current page
   */
  pageItems: GridItemData[]
  /**
   * A collection of all the items available
   */
  allItems: GridItemData[]
  /**
   * Current sort key for list view
   */
  sortBy: ResultsSortBy
}

type ResultsStoreAction = {
  /**
   * Sets the current page
   */
  setPage: (page: ResultsStoreState['page']) => void
  /**
   * Sets the count of items on each page
   */
  setPageSize: (pageSize: ResultsStoreState['pageSize']) => void
  /**
   * Sets the sort key for the results list
   */
  setSortBy: (sortBy: ResultsSortBy) => void
}

export type ResultsStore = ResultsStoreState & ResultsStoreAction

interface ResultsProviderProps extends PropsWithChildren {
  allItems: ResultsStoreState['allItems']
}

const ResultsContext = createContext<
  StoreApi<ResultsStore> | undefined
>(undefined)

export const ResultsProvider = ({
  children,
  allItems,
}: ResultsProviderProps) => {
  const navigate = useNavigate()
  const [searchParams, setSearchParams] = useSearchParams()
  const { pathname } = useLocation()

  const latestSearchParamsRef = useRef(searchParams)
  useEffect(() => {
    latestSearchParamsRef.current = searchParams
  }, [searchParams])

  const resultsStore = useRef(
    createStore<ResultsStore>((_, get) => ({
      page: Number(searchParams.get('page') ?? '1'),
      totalPages: 1,
      pageSize: Number(
        searchParams.get('pageSize') ??
          PAGE_SIZE_OPTIONS[0].toString(),
      ) as PageSizeOption,
      pageItems: [],
      allItems,
      sortBy: (() => {
        const raw = (searchParams.get('sort') ??
          'alphabetical') as ResultsSortBy
        const valid: ResultsSortBy[] = [
          'alphabetical',
          'version',
          'dependencyType',
          'dependents',
          'moduleType',
          'overallScore',
        ]
        return valid.includes(raw) ? raw : 'alphabetical'
      })(),
      setPage: (page: ResultsStoreState['page']) => {
        const next = new URLSearchParams(
          latestSearchParamsRef.current,
        )
        next.set('page', String(page))
        // keep current pageSize; if missing, use the store's current value
        if (!next.get('pageSize')) {
          next.set('pageSize', String(get().pageSize))
        }
        setSearchParams(next, { replace: true })
      },
      setPageSize: (pageSize: ResultsStoreState['pageSize']) => {
        const next = new URLSearchParams(
          latestSearchParamsRef.current,
        )
        next.set('pageSize', String(pageSize))
        // reset page to 1 when page size changes
        next.set('page', '1')
        setSearchParams(next, { replace: true })
      },
      setSortBy: (sortBy: ResultsSortBy) => {
        const next = new URLSearchParams(
          latestSearchParamsRef.current,
        )
        next.set('sort', sortBy)
        setSearchParams(next, { replace: true })
      },
    })),
  ).current

  /**
   * When new `items` are available: we assume there are results,
   * and we navigate to `.../<encodedUrlQuery>?page=1`
   */
  useEffect(() => {
    // only run this logic when we actually have a result set that implies “list view”
    // i.e., more than one item. Skip while loading (length === 0) and skip the single-item case.
    if (allItems.length > 1) {
      // remove trailing "/overview" (with or without trailing slash)
      const newPath = pathname.replace(/\/overview\/?$/, '')

      // only navigate if path is actually changing
      if (newPath !== pathname) {
        void navigate(newPath, { replace: true }) // avoid history spam
      }

      // ensure page=1 is present; clone instead of mutating in place
      if (!searchParams.get('page')) {
        const next = new URLSearchParams(searchParams)
        next.set('page', '1')
        next.set('pageSize', PAGE_SIZE_OPTIONS[0].toString())
        setSearchParams(next, { replace: true })
      }
    }
  }, [allItems, pathname, navigate, searchParams, setSearchParams])

  // keep store in sync with URL and data; clamp and validate values
  useEffect(() => {
    const defaultPageSize = PAGE_SIZE_OPTIONS[0]
    const rawPageSize = Number(
      searchParams.get('pageSize') ?? String(defaultPageSize),
    )
    const validPageSize =
      (
        (PAGE_SIZE_OPTIONS as unknown as number[]).includes(
          rawPageSize,
        )
      ) ?
        rawPageSize
      : defaultPageSize

    const totalPages = Math.max(
      1,
      Math.ceil(allItems.length / Math.max(1, validPageSize)),
    )

    const rawPage = Number(searchParams.get('page') ?? '1')
    const clampedPage = Math.min(Math.max(rawPage, 1), totalPages)

    const rawSort = (searchParams.get('sort') ??
      'alphabetical') as ResultsSortBy
    const validSorts: ResultsSortBy[] = [
      'alphabetical',
      'version',
      'dependencyType',
      'dependents',
      'moduleType',
      'overallScore',
    ]
    const sortBy =
      validSorts.includes(rawSort) ? rawSort : 'alphabetical'

    const start = (clampedPage - 1) * validPageSize
    const pageSlice = allItems.slice(start, start + validPageSize)

    const compareString = (x?: string, y?: string) =>
      (x ?? '').localeCompare(y ?? '', undefined, {
        numeric: true,
        sensitivity: 'base',
      })
    const compareNumber = (
      x?: number | null,
      y?: number | null,
      desc = false,
    ) => {
      const nx = x ?? -Infinity
      const ny = y ?? -Infinity
      return desc ? ny - nx : nx - ny
    }

    const pageItems = [...pageSlice].sort((a, b) => {
      switch (sortBy) {
        case 'alphabetical':
          return compareString(a.name, b.name)
        case 'version': {
          const va = a.version
          const vb = b.version
          return compareString(va, vb)
        }
        case 'dependencyType':
          return compareString(a.type, b.type)
        case 'dependents':
          return compareNumber(a.size, b.size, true)
        case 'moduleType': {
          const ma = a.to?.manifest?.type ?? ''
          const mb = b.to?.manifest?.type ?? ''
          return compareString(ma, mb)
        }
        case 'overallScore': {
          const scoreA = a.to?.insights.score
          const scoreB = b.to?.insights.score
          const sa = scoreA ? scoreA.overall : null
          const sb = scoreB ? scoreB.overall : null
          return compareNumber(sa, sb, true)
        }
      }
    })

    // reflect clamped/validated values in URL if needed
    if (
      rawPage !== clampedPage ||
      rawPageSize !== validPageSize ||
      rawSort !== sortBy
    ) {
      const next = new URLSearchParams(searchParams)
      next.set('page', String(clampedPage))
      next.set('pageSize', String(validPageSize))
      next.set('sort', sortBy)
      setSearchParams(next, { replace: true })
    }

    resultsStore.setState({
      page: clampedPage,
      pageSize: validPageSize as PageSizeOption,
      totalPages,
      pageItems,
      allItems,
      sortBy,
    })
  }, [allItems, searchParams, resultsStore, setSearchParams])

  return (
    <ResultsContext.Provider value={resultsStore}>
      {children}
    </ResultsContext.Provider>
  )
}

export const useResultsStore = <T,>(
  selector: (state: ResultsStore) => T,
) => {
  const store = useContext(ResultsContext)
  if (!store) {
    throw new Error(
      'useResultsStore must be used within a ResultsProvider',
    )
  }
  return useStore(store, selector)
}
