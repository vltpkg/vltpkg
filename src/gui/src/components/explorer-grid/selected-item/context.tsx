import { createContext, useContext, useRef } from 'react'
import { useStore, createStore } from 'zustand'
import { hydrate } from '@vltpkg/dep-id/browser'
import { useGraphStore } from '@/state/index.js'
import { fetchDetails } from '@/lib/external-info.js'

import { SOCKET_SECURITY_DETAILS } from '@/lib/constants/socket.js'

import type { StoreApi } from 'zustand'
import type { GridItemData } from '@/components/explorer-grid/types.js'
import type { DetailsInfo } from '@/lib/external-info.js'
import type { Insights } from '@vltpkg/query'
import type { PropsWithChildren } from 'react'
import type { SocketSecurityDetails } from '@/lib/constants/socket.js'
import type { PackageScore } from '@vltpkg/security-archive'
import type { Manifest } from '@vltpkg/types'

export type Tab = 'overview' | 'insights' | 'versions' | 'manifest'

const getSocketInsights = (
  insights?: Insights,
): SocketSecurityDetails[] | undefined => {
  if (!insights?.scanned) return

  return Object.entries(insights)
    .flatMap(([key, value]) => {
      if (typeof value === 'boolean' && value) {
        return SOCKET_SECURITY_DETAILS[
          key as keyof typeof SOCKET_SECURITY_DETAILS
        ] as SocketSecurityDetails
      }

      if (typeof value === 'object') {
        return Object.entries(value)
          .filter(([, subValue]) => subValue === true)
          .map(([subKey]) => {
            const parentInsight =
              SOCKET_SECURITY_DETAILS[
                key as keyof typeof SOCKET_SECURITY_DETAILS
              ]
            return (
                parentInsight && typeof parentInsight === 'object'
              ) ?
                (parentInsight[
                  subKey as keyof typeof parentInsight
                ] as SocketSecurityDetails)
              : undefined
          })
          .filter(Boolean)
      }

      return []
    })
    .filter(Boolean) as SocketSecurityDetails[]
}

type SelectedItemStoreState = DetailsInfo & {
  selectedItem: GridItemData
  activeTab: Tab
  manifest: Manifest | null
  packageScore?: PackageScore
  insights: SocketSecurityDetails[] | undefined
}

type SelectedItemStoreAction = {
  setActiveTab: (tab: SelectedItemStoreState['activeTab']) => void
}

export type SelectedItemStore = SelectedItemStoreState &
  SelectedItemStoreAction

const SelectedItemContext = createContext<
  StoreApi<SelectedItemStore> | undefined
>(undefined)

type SelectedItemProviderProps = PropsWithChildren & {
  selectedItem: GridItemData
}

export const SelectedItemProvider = ({
  children,
  selectedItem,
}: SelectedItemProviderProps) => {
  const specOptions = useGraphStore(state => state.specOptions)

  /**
   * We initialize the zustand store as a scoped state within the context:
   *
   * This brings Zustand into the react lifecycle and allows us top
   * initialize the store with the selected item while also limiting the access
   * to the store within the context's component tree.
   */
  const selectedItemStore = useRef(
    createStore<SelectedItemStore>(set => ({
      selectedItem,
      manifest: selectedItem.to?.manifest ?? null,
      packageScore: selectedItem.to?.insights.score,
      insights: getSocketInsights(selectedItem.to?.insights),
      activeTab: 'overview',
      setActiveTab: (
        activeTab: SelectedItemStoreState['activeTab'],
      ) => set(() => ({ activeTab })),
    })),
  ).current

  const fetchDetailsAsync = async (
    store: StoreApi<SelectedItemStore>,
  ) => {
    const state = store.getState()
    const item = state.selectedItem

    if (!item.to?.name) return

    const depIdSpec = hydrate(item.to.id, item.to.name, specOptions)
    const manifest = item.to.manifest ?? {}
    const abortController = new AbortController()

    try {
      for await (const d of fetchDetails(
        depIdSpec,
        abortController.signal,
        manifest,
      )) {
        store.setState(state => ({
          ...state,
          ...d,
        }))
      }
    } finally {
      abortController.abort()
    }
  }

  void fetchDetailsAsync(selectedItemStore)

  return (
    <SelectedItemContext.Provider value={selectedItemStore}>
      {children}
    </SelectedItemContext.Provider>
  )
}

export const useSelectedItemStore = <T,>(
  selector: (state: SelectedItemStore) => T,
) => {
  const store = useContext(SelectedItemContext)
  if (!store) {
    throw new Error(
      'useSelectedItemStore must be used within a SelectedItemProvider',
    )
  }
  return useStore(store, selector)
}
