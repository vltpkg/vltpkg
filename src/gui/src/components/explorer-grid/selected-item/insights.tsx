import { useSelectedItem } from '@/components/explorer-grid/selected-item/context.jsx'
import {
  getAlertColor,
  InsightBadge,
} from '@/components/explorer-grid/selected-item/insight-badge.jsx'

export const Insights = () => {
  const { insights } = useSelectedItem()

  if (!insights || insights.length === 0) return null

  return (
    <section className="flex w-full flex-wrap gap-3 px-6 pb-6">
      {insights.map((insight, idx) => (
        <InsightBadge
          key={idx}
          color={getAlertColor(insight.severity)}
          tooltipContent={`${insight.severity} severity`}>
          {insight.selector}
        </InsightBadge>
      ))}
    </section>
  )
}
