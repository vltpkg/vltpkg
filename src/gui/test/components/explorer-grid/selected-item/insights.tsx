import { test, expect, vi, beforeEach, afterEach } from 'vitest'
import { cleanup, render } from '@testing-library/react'
import html from 'diffable-html'
import { Insights } from '@/components/explorer-grid/selected-item/insights.jsx'
import { useGraphStore as useStore } from '@/state/index.js'
import { useSelectedItem } from '@/components/explorer-grid/selected-item/context.jsx'
import {
  SELECTED_ITEM,
  SELECTED_ITEM_DETAILS,
} from './__fixtures__/item.ts'
import type { SocketSecurityDetails } from '@/lib/constants/socket.js'

vi.mock(
  '@/components/explorer-grid/selected-item/context.jsx',
  () => ({
    useSelectedItem: vi.fn(),
  }),
)

vi.mock(
  '@/components/explorer-grid/selected-item/insight-badge.jsx',
  async () => {
    const actual = await import(
      '@/components/explorer-grid/selected-item/insight-badge.jsx'
    )
    return {
      ...actual,
      InsightBadge: 'gui-insight-badge',
    }
  },
)

expect.addSnapshotSerializer({
  serialize: v => html(v),
  test: () => true,
})

beforeEach(() => {
  vi.clearAllMocks()

  vi.mocked(useSelectedItem).mockReturnValue({
    selectedItem: SELECTED_ITEM,
    selectedItemDetails: SELECTED_ITEM_DETAILS,
    insights: undefined,
    activeTab: 'insight',
    setActiveTab: vi.fn(),
  })
})

afterEach(() => {
  const CleanUp = () => (useStore(state => state.reset)(), '')
  render(<CleanUp />)
  cleanup()
})

test('Insights renders default', () => {
  const Container = () => {
    return <Insights />
  }

  const { container } = render(<Container />)
  expect(container.innerHTML).toBe('')
})

test('Insights renders with socket insights', () => {
  const mockInsights: SocketSecurityDetails[] = [
    {
      selector: ':mock-selector',
      description: 'Mock description',
      category: 'Supply Chain',
      severity: 'medium',
    },
  ]

  vi.mocked(useSelectedItem).mockReturnValue({
    selectedItem: SELECTED_ITEM,
    selectedItemDetails: SELECTED_ITEM_DETAILS,
    insights: mockInsights,
    activeTab: 'insight',
    setActiveTab: vi.fn(),
  })

  const { container } = render(<Insights />)
  expect(container.innerHTML).toMatchSnapshot()
})
