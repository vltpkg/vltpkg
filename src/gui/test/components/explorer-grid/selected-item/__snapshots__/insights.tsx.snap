// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Insights renders with socket insights 1`] = `

<section class="flex w-full flex-wrap gap-3 px-6 pb-6">
  <gui-insight-badge
    color="yellow"
    tooltipcontent="medium severity"
  >
    :mock-selector
  </gui-insight-badge>
</section>

`;
